cmd/metal-api/internal/auditing/auditing-interceptor.go

```diff
@@ -0,0 +1,205 @@
+		}
+		requestID := uuid.New().String()
+		childCtx := context.WithValue(ctx, rest.RequestIDKey, requestID)
+
+		auditReqContext := []any{
+			"rqid", requestID,
+			"method", info.FullMethod,
+			"kind", "grpc",
+		}
+		user := security.GetUserFromContext(ctx)
+		if user != nil {
+			auditReqContext = append(
+				auditReqContext,
+				"user", user.EMail,
+				"tenant", user.Tenant,
+			)
+		}
+		err = a.Index(auditReqContext...)
+		if err != nil {
+			return nil, err
+		}
+		resp, err = handler(childCtx, req)
+		if err != nil {
+			auditRespContext := append(auditReqContext, "err", err)
+			err2 := a.Index(auditRespContext...)
+			if err2 != nil {
+				logger.Errorf("unable to index error: %v", err2)
+			}
+			return nil, err
+		}
+		auditRespContext := append(auditReqContext, "resp", resp)
+		err = a.Index(auditRespContext...)
+		return resp, err
+	}
+}
+
+func StreamServerInterceptor(a Auditing, logger *zap.SugaredLogger, shouldAudit func(fullMethod string) bool) grpc.StreamServerInterceptor {
+	if a == nil {
+		logger.Fatal("cannot use nil auditing to create stream server interceptor")
+	}
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		if !shouldAudit(info.FullMethod) {
+			return handler(srv, ss)
+		}
+		requestID := uuid.New().String()
+		auditReqContext := []any{
+			"rqid", requestID,
+			"method", info.FullMethod,
+			"kind", "grpc-stream",
+		}
+
+		user := security.GetUserFromContext(ss.Context())
+		if user != nil {
+			auditReqContext = append(
+				auditReqContext,
+				"user", user.EMail,
+				"tenant", user.Tenant,
+			)
+		}
+		err := a.Index(auditReqContext...)
+		if err != nil {
+			return err
+		}
+		err = handler(srv, ss)
+		if err != nil {
+			auditRespContext := append(auditReqContext, "err", err)
+			err2 := a.Index(auditRespContext...)
+			if err2 != nil {
+				logger.Errorf("unable to index error: %v", err2)
+			}
+			return err
+		}
+		auditRespContext := append(auditReqContext, "finished", true)
+		err = a.Index(auditRespContext...)
+		return err
+	}
+}
+
+func HttpFilter(a Auditing, logger *zap.SugaredLogger) restful.FilterFunction {
+	if a == nil {
+		logger.Fatal("cannot use nil auditing to create http middleware")
+	}
+	return func(request *restful.Request, response *restful.Response, chain *restful.FilterChain) {
+		r := request.Request
+
+		switch r.Method {
+		case http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete:
+			break
+		default:
+			chain.ProcessFilter(request, response)
+			return
+		}
+
+		excluded, ok := request.SelectedRoute().Metadata()[Exclude].(bool)
+		if ok && excluded {
+			logger.Debugw("excluded route from auditing through metadata annotation", "path", request.SelectedRoute().Path())
+			chain.ProcessFilter(request, response)
+			return
+		}
+
+		requestID := r.Context().Value(rest.RequestIDKey)
+		auditReqContext := []any{
+			"rqid", requestID,
+			"method", r.Method,
+			"path", r.URL.Path,
```

Comment:
What do you think about adding `request.Request.RemoteAddr` and `request.HeaderParameter("x-forwarded-for")` for tracing the request origin?

Reply:
Good point. Might be interesting. I'll give it a try.